File: FindPeakElementInArray.java
public class FindPeakElementInArray {
    public static int findPeakElement(int[] nums) {
        // int max = nums[0], index = 0;
        // for(int i=1; i<nums.length; i++){
        //     System.out.println(nums[i]);
        //     if(i+1 < nums.length && max < nums[i] && nums[i-1] < nums[i] && nums[i] > nums[i+1]){
        //         max = nums[i];
        //         index = i;
        //     }
        //     else if(i+1 == nums.length && nums[i] > max){
        //         max = nums[i];
        //         index = i;
        //     }
        // }
        return bs(nums, 0, nums.length-1);
    }
    
    public static int bs(int[] nums, int l, int r){
        int m = (l+r)/2;
        if (l == r) 
            return m; 
        if (nums[m] > nums[m+1]) 
            return bs(nums, l, m); 
        return bs(nums, m+1, r); 
        
    }
    public static void main(String[] args) {
        int[] nums = {1,1,3,4,5,6};
        System.out.println(findPeakElement(nums));
    }
}

File: FindPeakElementInArray.py
# Time Complexity : O(log n) 45ms //Based on leetcode runtime
#  Space Complexity : O(1) 14.1 //Based on leetcode
#  Did this code successfully run on Leetcode : Yes
#  Any problem you faced while coding this : N/A

#  Your code here along with comments explaining your approach

class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        #If no elments in nums return -1, if only one elment present the return 0 index
        n = len(nums)
        if not nums:
            return -1
        if n == 1:
            return 0
        
        #Defining lower and upper bound
        left = 0
        right = n - 1
        
        #Binary Search on nums till we find the desire output
        while left <= right:
            mid = (left+right) // 2 
            
            #If the mid is less than the upper bound and the current number is greater that the mid + 1 then make right provisonal peak
            if mid < right and nums[mid] > nums[mid + 1]:
                right = mid
            #Else if element at mid is less than the element next to it then make the lower bound as that and again run the loop
            elif mid < right and nums[mid] < nums[mid + 1]:
                left = mid + 1
            else:
                return mid
        return right